syft/pkg/cataloger/nix/cataloger.go

```diff
@@ -0,0 +1,101 @@
+package nix
+
+import (
+	"fmt"
+
+	"github.com/bmatcuk/doublestar/v4"
+
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/source"
+)
+
+const (
+	catalogerName = "nix-store-cataloger"
+	nixStoreGlob  = "**/nix/store/*"
+)
+
+type Cataloger struct{}
+
+func NewStoreCataloger() *Cataloger {
+	return &Cataloger{}
+}
+
+func (c *Cataloger) Name() string {
+	return catalogerName
+}
+
+func (c *Cataloger) Catalog(resolver source.FileResolver) ([]pkg.Package, []artifact.Relationship, error) {
+	// we want to search for only directories, which isn't possible via the stereoscope API, so we need to apply the glob manually on all returned paths
+	var pkgs []pkg.Package
+	var filesByPath = make(map[string]*source.LocationSet)
+	for location := range resolver.AllLocations() {
+		matchesStorePath, err := doublestar.Match(nixStoreGlob, location.RealPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to match nix store path: %w", err)
+		}
+
+		parentStorePath := findParentNixStorePath(location.RealPath)
+		if parentStorePath != "" {
+			if _, ok := filesByPath[parentStorePath]; !ok {
+				s := source.NewLocationSet()
+				filesByPath[parentStorePath] = &s
+			}
+			filesByPath[parentStorePath].Add(location)
+		}
+
+		if !matchesStorePath {
```

Comment:
shouldn't this happen before `s := source.NewLocationSet() ... filesByPath[parentStorePath] = &s`? and this would go away if using `FilesByGlob`?

Reply:
>  and this would go away if using FilesByGlob:

the search is for the `/nix/store/<hash>-<name>-<version>-<output>` directory, not a file. We are collecting locations that belong to a particular store path within the same loop that we are looking for store paths to not make redundant request: why ask for `/nix/store/PATH/**` when I've already seen all of the locations in an earlier loop?

> shouldn't this happen before s := source.NewLocationSet() ... filesByPath[parentStorePath] = &s

I don't think so; we want to capture:
a) `/nix/store/PATH` : the location to capture as a package
b) `/nix/store/PATH/**`: all files/directories that are "owned" by that package

We first check if the path is a parent of a store path (e.g. when cataloging `/nix/store/PATH/a/b/c` it's parent store path is `/nix/store/PATH`). We only continue forth if we have a new store path to catalog, in which case we need to parse the directory name to understand what the package is.

Reply:
Right, sorry I understand now about capturing directories